pkg/nsx/services/subnet/subnet.go

```diff
@@ -0,0 +1,241 @@
+func (service *SubnetService) CreateOrUpdateSubnet(obj client.Object, vpcInfo *common.VPCResourceInfo, tags []model.Tag) (string, error) {
+	value, ok := obj.GetAnnotations()[common.TagScopeSubnetType]
+	if ok {
+		tags = append(tags, model.Tag{
+			Scope: String(common.TagScopeSubnetType),
+			Tag:   String(value),
+		})
+	}
+	uid := string(obj.GetUID())
+	nsxSubnet, err := service.buildSubnet(obj, tags)
+	if err != nil {
+		log.Error(err, "failed to build Subnet")
+		return "", err
+	}
+	switch obj.(type) {
+	case *v1alpha1.Subnet:
+		nsxSubnets := service.SubnetStore.GetByIndex(common.TagScopeSubnetCRUID, uid)
+		changed := false
+		if len(nsxSubnets) == 0 {
+			changed = true
+		} else {
+			changed = common.CompareResource(SubnetToComparable(&nsxSubnets[0]), SubnetToComparable(nsxSubnet))
+		}
+		if !changed {
+			log.Info("subnet not changed, skip updating", "subnet.Id", *nsxSubnets[0].Id)
+			return *nsxSubnets[0].Id, nil
+		}
+	}
+	orgRoot, err := service.WrapHierarchySubnet(nsxSubnet, vpcInfo)
+	if err != nil {
+		log.Error(err, "WrapHierarchySubnet failed")
+		return "", err
+	}
+	if err = service.NSXClient.OrgRootClient.Patch(*orgRoot, &EnforceRevisionCheckParam); err != nil {
+		return "", err
+	}
+	realizeService := realizestate.InitializeRealizeState(service.Service)
+	if err = realizeService.CheckRealizeState(retry.DefaultRetry, *nsxSubnet.Path, "RealizedLogicalSwitch"); err != nil {
+		return "", err
+	}
+	// Get Subnet from NSX after patch operation as NSX renders several fields like `path`/`parent_path`.
+	if *nsxSubnet, err = service.NSXClient.SubnetsClient.Get(vpcInfo.OrgID, vpcInfo.ProjectID, vpcInfo.VPCID, *nsxSubnet.Id); err != nil {
+		return "", err
+	}
+	if err = service.SubnetStore.Operate(nsxSubnet); err != nil {
+		return "", err
+	}
+	log.Info("successfully updated nsxSubnet", "nsxSubnet", nsxSubnet)
+	return *nsxSubnet.Id, nil
+}
+
+func (service *SubnetService) DeleteSubnet(obj client.Object, vpcInfo *common.VPCResourceInfo) error {
+	nsxSubnets := service.SubnetStore.GetByIndex(common.TagScopeSubnetCRUID, string(obj.GetUID()))
+	for _, nsxSubnet := range nsxSubnets {
+		nsxSubnet.MarkedForDelete = &MarkedForDelete
+		// WrapHighLevelSubnet will modify the input subnet, make a copy for the following store update.
+		subnetCopy := nsxSubnet
+		orgRoot, err := service.WrapHierarchySubnet(&nsxSubnet, vpcInfo)
+		if err != nil {
+			return err
+		}
+		if err = service.NSXClient.OrgRootClient.Patch(*orgRoot, &EnforceRevisionCheckParam); err != nil {
+			// Subnets that are not deleted successfully will finally be deleted by GC.
+			log.Error(err, "failed to delete Subnet", "ID", *nsxSubnet.Id)
+			continue
+		}
+		if err = service.SubnetStore.Operate(&subnetCopy); err != nil {
+			return err
+		}
+		log.Info("successfully deleted nsxSubnet", "nsxSubnet", nsxSubnet)
+	}
+	return nil
+}
+
+func (service *SubnetService) DeleteIPAllocation(orgID, projectID, vpcID, subnetID string) error {
```

Comment:
In which case this function will be used?

Reply:
IP allocation will block deletion of subnet, so in previous implementation, nsx-operator delete ip allocation before deleing subnet. But @lxiaopei mentioned that it's ok when we hit deletion error when deleting subnet with allocation, so the function is not used for now.

But I think in GC, we may need it to ensure subnet can be deleted, so I keep the function here.